stablehlo/transforms/StablehloWrapInComposite.cpp

```diff
@@ -0,0 +1,216 @@
+  builder.createBlock(&implFunc.getBody(), implFunc.begin(), argTypes, argLocs);
+
+  // Map the operands of the original op to the arguments of the newly created
+  // function.
+  for (const auto& operand : llvm::enumerate(implOp->getOperands())) {
+    mapping.map(operand.value(), implFunc.getArgument(operand.index()));
+  }
+
+  // Clone the original operation into the body of the new function,
+  // using the value mapping to remap operands.
+  mlir::Operation* cloned_op = builder.clone(*implOp, mapping);
+
+  // Create the return operation, returning the results of the cloned op.
+  llvm::SmallVector<mlir::Value> results;
+  results.append(cloned_op->getResults().begin(),
+                 cloned_op->getResults().end());
+  builder.create<mlir::func::ReturnOp>(implFunc.getBody().getLoc(), results);
+
+  // Add the newly created function to the module's symbol table and make it
+  // private.
+  mlir::SymbolTable symbol_table(module);
+  implFunc.setPrivate();
+  symbol_table.insert(implFunc);
+
+  return implFunc;
+}
+
+// A ConversionPattern that matches any operation and rewrites it as a
+// stablehlo::CompositeOp. The original operation's functionality is
+// encapsulated within a newly created private function.
+class ConvertGenericOp : public ConversionPattern {
+ public:
+  explicit ConvertGenericOp(MLIRContext* ctx)
+      : ConversionPattern(MatchAnyOpTypeTag(), /*benefit=*/1, ctx) {}
+
+  LogicalResult matchAndRewrite(
+      Operation* op, ArrayRef<Value> operands,
+      ConversionPatternRewriter& rewriter) const override {
+    // get the enclosing module
+    auto module = op->getParentOfType<ModuleOp>();
+    MLIRContext* context = op->getContext();
+    if (module == nullptr) {
+      return rewriter.notifyMatchFailure(op, "Failed to find enclosing module");
+    }
+    auto implFunc = buildStableHLOCompositeImplFunc(module, op);
+    auto name = op->getName().getStringRef();
+
+    llvm::SmallVector<mlir::Value> compositeOperands(op->operand_begin(),
+                                                     op->operand_end());
+    auto compositeOp = rewriter.create<stablehlo::CompositeOp>(
+        op->getLoc(), op->getResultTypes(), compositeOperands, name,
+        DictionaryAttr::get(context, op->getAttrs()), implFunc.getSymName());
+    rewriter.replaceOp(op, compositeOp.getResults());
+    return success();
+  }
+};
+
+class StablehloWrapInCompositePass
+    : public impl::StablehloWrapInCompositePassBase<
+          StablehloWrapInCompositePass> {
+ public:
+  StablehloWrapInCompositePass()
+      : StablehloWrapInCompositePassBase<StablehloWrapInCompositePass>() {}
+  StablehloWrapInCompositePass(const StablehloWrapInCompositePassOptions& opts)
+      : StablehloWrapInCompositePassBase<StablehloWrapInCompositePass>(opts) {}
+  explicit StablehloWrapInCompositePass(
+      std::function<bool(Operation*)> opPredicate) {
+    this->opPredicate = opPredicate;
```

Comment:
Can we use list init instead? Also shouldn't we call the pass base here as well?

`: StablehloWrapInCompositePassBase<StablehloWrapInCompositePass>(opts), opPredicate(opPredicate)`